plugins/plugin-github/codenvy-plugin-github-factory-resolver/src/main/java/com/codenvy/plugin/factory/github/factory/resolver/GithubFactoryParametersResolver.java

```diff
@@ -0,0 +1,186 @@
+/*
+ *  [2012] - [2016] Codenvy, S.A.
+ *  All Rights Reserved.
+ *
+ * NOTICE:  All information contained herein is, and remains
+ * the property of Codenvy S.A. and its suppliers,
+ * if any.  The intellectual and technical concepts contained
+ * herein are proprietary to Codenvy S.A.
+ * and its suppliers and may be covered by U.S. and Foreign Patents,
+ * patents in process, and are protected by trade secret or copyright law.
+ * Dissemination of this information or reproduction of this material
+ * is strictly forbidden unless prior written permission is obtained
+ * from Codenvy S.A..
+ */
+package com.codenvy.plugin.factory.github.factory.resolver;
+
+import org.eclipse.che.api.core.BadRequestException;
+import org.eclipse.che.api.factory.server.FactoryParametersResolver;
+import org.eclipse.che.api.factory.shared.dto.Factory;
+import org.eclipse.che.api.machine.shared.dto.LimitsDto;
+import org.eclipse.che.api.machine.shared.dto.MachineConfigDto;
+import org.eclipse.che.api.machine.shared.dto.MachineSourceDto;
+import org.eclipse.che.api.workspace.shared.dto.EnvironmentDto;
+import org.eclipse.che.api.workspace.shared.dto.ProjectConfigDto;
+import org.eclipse.che.api.workspace.shared.dto.SourceStorageDto;
+import org.eclipse.che.api.workspace.shared.dto.WorkspaceConfigDto;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.validation.constraints.NotNull;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static java.util.Collections.singletonList;
+import static org.eclipse.che.dto.server.DtoFactory.newDto;
+
+/**
+ * Provides Factory Parameters resolver for github repositories.
+ *
+ * @author Florent Benoit
+ */
+public class GithubFactoryParametersResolver implements FactoryParametersResolver {
+
+    /**
+     * Logger.
+     */
+    private static final Logger LOG = LoggerFactory.getLogger(GithubFactoryParametersResolver.class);
+
+    /**
+     * Parameter name.
+     */
+    protected static final String URL_PARAMETER_NAME = "url";
+
+    /**
+     * Connection timeout of 10seconds.
+     */
+    private static final int CONNECTION_TIMEOUT = (int) TimeUnit.SECONDS.toMillis(10);
+
+    /**
+     * Regexp to find repository details (repository name, project name and branch and subfolder)
+     * Examples of valid URLs are in the test class.
+     */
+    private static final Pattern GITHUB_PATTERN = Pattern.compile("^(?:http)(?:s)?(?:\\:\\/\\/)github.com/(?<repoUser>[^/]++)/(?<repoName>[^/]++)(?:/tree/(?<branchName>[^/]++)(?:/(?<subFolder>[^/]++))?)?$");
```

Comment:
Look like next urls are invalid for that regexp. Is it supposed that next urls are valid?
https://github.com/eclipse/che/tree/master/
https://github.com/eclipse/che/tree/master/dashboard/
https://github.com/eclipse/che/tree/master/plugins/plugin-git/che-plugin-git-ext-git
https://github.com/eclipse/che/tree/master/plugins/plugin-git/che-plugin-git-ext-git/